primitives/orderbook/src/types.rs

```diff
@@ -0,0 +1,595 @@
+	/// if returned is false then underflow occurred during division
+	pub fn update_avg_price_and_filled_qty(&mut self, price: Decimal, amount: Decimal) -> bool {
+		let mut temp = self.avg_filled_price.saturating_mul(self.filled_quantity);
+		temp = temp.saturating_add(amount.saturating_mul(price));
+		self.filled_quantity = self.filled_quantity.saturating_add(amount);
+		println!("self.filled_quantity: {:?}\ntemp: {:?}", self.filled_quantity, temp);
+		match temp.checked_div(self.filled_quantity) {
+			Some(quotient) => {
+				println!("Quotient: {:?}", quotient);
+				self.avg_filled_price = quotient;
+				true
+			},
+			None => false,
+		}
+	}
+
+	pub fn available_volume(&self, other_price: Option<Decimal>) -> Decimal {
+		//this if for market bid order
+		if self.qty.is_zero() {
+			println!(
+				"quote_order_qty: {:?}, avg_filled_price: {:?}, filled_quantity: {:?}",
+				self.quote_order_qty, self.avg_filled_price, self.filled_quantity
+			);
+			return rounding_off(
+				self.quote_order_qty
+					.saturating_sub(self.avg_filled_price.saturating_mul(self.filled_quantity)),
+			)
+		}
+		//this is for market ask order
+		if self.order_type == OrderType::MARKET {
+			rounding_off(
+				self.qty
+					.saturating_sub(self.filled_quantity)
+					.saturating_mul(other_price.unwrap_or_default()),
+			)
+		}
+		//this is for limit orders
+		else {
+			// We cannot use avg. price here as limit orders might not have avg_price defined
+			// if they are not yet matched and just inserted into the book
+			rounding_off(self.qty.saturating_sub(self.filled_quantity).saturating_mul(self.price))
+		}
+	}
+	// TODO: how to gate this only for testing
+	#[cfg(feature = "std")]
+	pub fn random_order_for_testing(
+		pair: TradingPair,
+		side: OrderSide,
+		order_type: OrderType,
+	) -> Self {
+		use rand::Rng;
+		let mut rng = rand::thread_rng();
+		Self {
+			stid: Default::default(),
+			client_order_id: H256([1u8; 32]),
+			avg_filled_price: Decimal::zero(),
+			fee: Decimal::zero(),
+			filled_quantity: Decimal::zero(),
+			status: OrderStatus::OPEN,
+			id: H256([2u8; 32]),
+			user: AccountId::new(rng.gen()),
+			main_account: AccountId::new([0u8; 32]),
+			pair,
+			side,
+			order_type,
+			qty: Decimal::from(rng.gen_range(MIN_QTY..MAX_QTY)),
+			price: Decimal::from(rng.gen_range(MIN_PRICE..MAX_PRICE)),
+			quote_order_qty: Decimal::zero(),
+			timestamp: 1,
+			overall_unreserved_volume: Decimal::zero(),
+		}
+	}
+}
+
+pub fn rounding_off(a: Decimal) -> Decimal {
+	// if we want to operate with a precision of 8 decimal places,
+	// all calculations should be done with latest 9 decimal places
+	a.round_dp_with_strategy(9, RoundingStrategy::ToZero)
+}
+
+#[cfg(test)]
+mod tests {
+	use crate::types::{ObMessage, UserActions};
+
+	#[test]
+	pub fn test_ob_message() {
+		let msg = ObMessage { stid: 0, action: UserActions::BlockImport(1) };
+
+		println!("OBMessage: {:?}", serde_json::to_string(&msg).unwrap());
```

Comment:
Do we need this print? why not used logger?

Reply:
Same as above